lib/controller/controller.py

```diff
@@ -78,6 +78,544 @@ def pause(self):
+                    self.close(
+                        "\nSkipped the target due to {0} status code".format(self.status_skip),
+                        skip=True
+                    )
+            break
+
+
+
+    # Preparation between subdirectory scans
+    def prepare(self):
+        while not self.directories.empty():
+            gc.collect()
+            self.current_job += 1
+            self.index = 0
+            self.current_directory = self.directories.get()
+            self.output.warning(
+                "[{1}] Starting: {0}".format(
+                    self.current_directory, time.strftime("%H:%M:%S")
+                )
+            )
+            self.fuzzer.requester.base_path = self.output.base_path = self.base_path + self.current_directory
+            self.fuzzer.start()
+            self.process_paths()
+
+        self.report.completed = True
+
+    # Add directory to the recursion queue
+    def add_directory(self, path):
+        dirs = []
+        added = False
+        path = path.split("?")[0].split("#")[0]
+        full_path = self.current_directory + path
+
+        if any([path.startswith(directory) for directory in self.exclude_subdirs]):
+            return False
+
+        # Avoid paths contain consecutive slashes, we haven't had good handler for it yet
+        if self.deep_recursive and "//" not in path:
+            for i in range(1, path.count("/")):
+                dir = self.current_directory + "/".join(path.split("/")[:i]) + "/"
+                dirs.append(dir)
+
+        if self.force_recursive:
+            if not full_path.endswith("/"):
+                full_path += "/"
+            dirs.append(full_path)
+        elif self.recursive and full_path.endswith("/"):
+            dirs.append(full_path)
+
+        for dir in list(set(dirs)):
+            if dir in self.pass_dirs:
+                continue
+            elif self.recursion_depth and dir.count("/") > self.recursion_depth:
+                continue
+
+            self.directories.put(dir)
+            self.pass_dirs.append(dir)
+
+            self.jobs_count += 1
+            added = True
+        return added
+
+    # Resolve the redirect and add the path to the recursion queue
+    # if it's a subdirectory of the current URL
+    def add_redirect_directory(self, path):
+        base_path = "/" + self.base_path + self.current_directory + path.path
+
+        redirect_url = urljoin(self.requester.base_url, path.redirect)
+        redirect_path = urlparse(redirect_url).path
+
+        if redirect_path == base_path + "/":
+            path = redirect_path[len(self.base_path + self.current_directory) + 1:]
+
+            return self.add_directory(path)
+
+        return False
+
+    def close(self, msg=None, skip=False):
+        self.fuzzer.stop()
+        self.output.error(msg)
+        if skip:
+            raise SkipTargetInterrupt
+
+        self.report_manager.update_report(self.report)
+        exit(0)
+
+
```

Comment:
There are too many new lines here, please follow pep8 rules: https://www.python.org/dev/peps/pep-0008/

Reply:
Btw, what is the difference between `Controller` and `LibController`?

Reply:
> Btw, what is the difference between `Controller` and `LibController`?

I think so, i can't remember